src/Access/AccessControlManager.cpp

```diff
@@ -32,6 +41,110 @@ namespace
+
+            if (enable_tls_lc_str == "starttls")
+                params.enable_tls = LDAPServerParams::TLSEnable::YES_STARTTLS;
+            else if (config.getBool(ldap_server_config + ".enable_tls"))
+                params.enable_tls = LDAPServerParams::TLSEnable::YES;
+            else
+                params.enable_tls = LDAPServerParams::TLSEnable::NO;
+        }
+
+        if (has_tls_cert_verify)
+        {
+            String tls_cert_verify_lc_str = config.getString(ldap_server_config + ".tls_cert_verify");
+            boost::to_lower(tls_cert_verify_lc_str);
+
+            if (tls_cert_verify_lc_str == "never")
+                params.tls_cert_verify = LDAPServerParams::TLSCertVerify::NEVER;
+            else if (tls_cert_verify_lc_str == "allow")
+                params.tls_cert_verify = LDAPServerParams::TLSCertVerify::ALLOW;
+            else if (tls_cert_verify_lc_str == "try")
+                params.tls_cert_verify = LDAPServerParams::TLSCertVerify::TRY;
+            else if (tls_cert_verify_lc_str == "demand")
+                params.tls_cert_verify = LDAPServerParams::TLSCertVerify::DEMAND;
+            else
+                throw Exception("Bad value for 'tls_cert_verify' entry, allowed values are: 'never', 'allow', 'try', 'demand'", ErrorCodes::BAD_ARGUMENTS);
+        }
+
+        if (has_ca_cert_dir)
+            params.ca_cert_dir = config.getString(ldap_server_config + ".ca_cert_dir");
+
+        if (has_ca_cert_file)
+            params.ca_cert_file = config.getString(ldap_server_config + ".ca_cert_file");
+
+        if (has_port)
+        {
+            const auto port = config.getInt64(ldap_server_config + ".port");
+            if (port < 0 || port > 65535)
+                throw Exception("Bad value for 'port' entry", ErrorCodes::BAD_ARGUMENTS);
+
+            params.port = port;
+        }
+        else
+            params.port = (params.enable_tls == LDAPServerParams::TLSEnable::YES ? 636 : 389);
+
+        return params;
+    }
+
+    void parseAndAddLDAPServers(ExternalAuthenticators & external_authenticators, const Poco::Util::AbstractConfiguration & config, Poco::Logger * log)
+    {
+        Poco::Util::AbstractConfiguration::Keys ldap_server_names;
+        config.keys("ldap_servers", ldap_server_names);
+
+        for (const auto & ldap_server_name : ldap_server_names)
+        {
+            try
+            {
+                external_authenticators.setLDAPServerParams(ldap_server_name, parseLDAPServer(config, ldap_server_name));
+            }
+            catch (...)
+            {
+                tryLogCurrentException(log, "Could not parse LDAP server " + backQuote(ldap_server_name));
+            }
+        }
+    }
+
+    auto parseExternalAuthenticators(const Poco::Util::AbstractConfiguration & config, Poco::Logger * log)
+    {
+        auto external_authenticators = std::make_unique<ExternalAuthenticators>();
+        parseAndAddLDAPServers(*external_authenticators, config, log);
```

Comment:
Don't we want to put ldap config outside as we plan to use it not only for authentication?

Reply:
No, because it logically belongs here, and it will be very ugly to make those things visible here, where they are used.